First pass at port API needed for port UI

alexr00 opened the pull request:
Part of https://github.com/microsoft/vscode-remote-release/issues/1777 and https://github.com/microsoft/vscode/issues/81388

roblourens commented:
Hey @Chuxel, we were talking about this and I wanted to see what you think - is there any reason to prefer ssh-forwarded ports over ports forwarded by vscode? 

The SSH extension can declare that it has already forwarded some ports (like ones declared in the SSH config file) but I don't think we need a way for the SSH extension to actually implement the port forward.

roblourens commented on src/vs/vscode.proposed.d.ts:
```diff
@@ -33,7 +33,37 @@ declare module 'vscode' {
+		 * Allows ports that have been forwarded in a workspace to be restored next time that workspace is opened.
+		 * Alternative to this: have API for getting all the ports then an event when they change so that the extension can save and restore ports(as it can do today).
+		 */
+		restorePorts: boolean;
```
Need to distinguish between the vscode server port which shouldn't be restored, and others. The alternative suggestion might be better so the extension can differentiate.

alexr00 replied:
From what I understand, the vscode server port _must_ be forwarded by the extension. This API deliberately doesn't provide a way to tell the core about that port, so it won't know about it to restore it.

What other ports shouldn't be restored?

alexandrudima replied:
IMHO `restorePorts: boolean` should be a user setting, not something a provider gets to decide.

fara-nak replied:
We were talking about this with Sri today, that if we switch to using this feature, how would you support the ports that already been shared in a session but because user closed the vso (hence vscode), the ports have been killed. so I assume this restorePorts would solve this issue, am I right? but then where do you save the forwarded port to forward them for next session?

alexr00 replied:
@fara-nak Instead of having `restorePorts` in the API, I'd like to make it a setting. So when the user opens a workspace again the ports are automatically re-forwarded.

Chuxel commented:
> Hey @Chuxel, we were talking about this and I wanted to see what you think - is there any reason to prefer ssh-forwarded ports over ports forwarded by vscode?
>
> The SSH extension can declare that it has already forwarded some ports (like ones declared in the SSH config file) but I don't think we need a way for the SSH extension to actually implement the port forward.

I *think* using a common mechanism should work.  

Playing this out a bit - The main kinds of things that come to mind are around proxies, authentication, etc, but at this point we should be already connected to the other side and we've already authenticated.

The SSH protocol itself allows you to do things like reverse forwarding and allowing connections from things other than localhost, but any dynamic code we have in place today doesn't support that anyway. That seems better to add advanced config to the actual SSH config file.

The one subtlety that bit us on Containers is that "publish" is different than an actual forward.  Our forwarding mechanism works even if a host only accepts connections from localhost while publishing via Docker will not (hosts have to use 0.0.0.0 or '*').  I don't think that applies to SSH given we forward localhost as I recall.

UX wise, it probably makes sense to distinguish between the ones forwarded via SSH and a common mechanism like we'll need to for containers just in case we need to explain it.

mjbvz commented on src/vs/vscode.proposed.d.ts:
```diff
@@ -33,7 +33,37 @@ declare module 'vscode' {
 
-	export type ResolverResult = ResolvedAuthority & ResolvedOptions;
+	export interface Port extends Disposable {
+		remotePort: number;
```
Not sure if this is will matter for remote or not but one thing for the VSO embedder is that the `localPort` would actually be `https` address instead of just a port number

alexandrudima replied:
On the web, there is no resolver executing, so VSO does not interact / will not interact with this API because they implement port forwarding through the embedder API, not through extension API.

alexr00 replied:
+1 for calling it `Tunnel` instead.

alexr00 commented on src/vs/vscode.proposed.d.ts:
```diff
@@ -33,7 +33,37 @@ declare module 'vscode' {
 	}
 
-	export type ResolverResult = ResolvedAuthority & ResolvedOptions;
+	export interface Port extends Disposable {
```
I don't think that's needed here. `Port`  is used for `candidates`, `published` (already un-mutable), and in the `Resolver` for `forwardPort`. `forwardPort` is called from the core for forwarding a port, so the core gets to decide if it's un-forwardable. 

It is needed in `toForward` though. Added.

alexr00 commented on src/vs/vscode.proposed.d.ts:
```diff
@@ -33,7 +33,37 @@ declare module 'vscode' {
+		 * An array of ports to be forwarded once connected. For example,
+		 * if the extension wants to have a configuration file with some default ports, they could be passed through for forwarding here.
+		 */
+		toForward?: { remote: number, local: number, name: string }[];
```
The properties are slightly different. I'm not sure that `Port` needs a name, because that's something only the user can provide (either through new UI or through a config file). I also added a `closeable` property here, which I don't see a need for on `Port` right now.

alexandrudima commented on src/vs/vscode.proposed.d.ts:
```diff
@@ -33,7 +33,38 @@ declare module 'vscode' {
+		 * An array of ports to be forwarded once connected. For example,
+		 * if the extension wants to have a configuration file with some default ports, they could be passed through for forwarding here.
+		 */
+		toForward?: { remote: number, local: number, name: string, closeable?: boolean }[];
```
What is the difference between `toForward` and `candidates` ? Why is here a type inlined instead of using `Port`. Perhaps `Port` should have the `closeable` property too.

alexr00 replied:
`candidates` are ports that the user might want to forward. We will show them in the port UI to make it easy/obvious that these might be useful ports. `toForward` are ports that the extension knows should definitely be forwarded. `toForward` ports will be forwarded.

alexr00 replied:
I don't think that `closeable` or `name` is appropriate in the other places. A `candidate` port doesn't need a name, because the user hasn't taken action on it yet. Same for a `published` port. Similarly, `candidate` and `published` ports aren't closeable. I'm also not sure that when an extension implements `forwardPort` they should be allowed to decide if that port that they're forwarding is `closeable`.

alexandrudima replied:
But then `toForward` should be removed and the extension can call in a loop `asExternalUri`.

alexr00 replied:
`asExternalUri` doesn't allow the extension to give the port a name or choose a specific port.

alexandrudima replied:
But then should we just add separate port forwarding API, i.e. acknowledge that `asExternalUri` is not sufficient and add something more powerful?

alexr00 replied:
I think the use of `asExternalUri` is different from `toForward`:
- Any extension can call `asExternalUri` to get a usable uri. Generally, I expect that extension calling `asExternalUri` only care that they have a usable uri, not the specifics of how they get that usable uri. We don't actually want them to care about specifics. They should "just work"
- Only a resolver can use `toForward` since it is used as part of the remote session setup. The session setup knows that it is remote, so it's the resolver's job to care about specifics such as port numbers.

alexandrudima replied:
Given the functionality desire that a resolver can forward ports, there are two ways to go about it:

1. Add `PortInformation.toForward` as is proposed here. This is an expansion of the current API. This means that only resolvers can forward ports, but then they cannot stop the forwarding because `toForward` is provided by them once and then resolvers cannot call `Dispose` on something to stop them. So if a config file changes, it is not possible for a resolver to react to it and stop forwarding the current ports and begin forwarding the new ports.

2. Add `vscode.workspace.forwardPort(...): Port`. This means we do not add `PortInformation.toForward`. This is also an expansion of the current API. But this means any extension, including resolvers, can forward ports. Extensions end up forwarding ports today only as a side-effect of calling `asExternalUri`, so extensions already have the "access" to do port forwarding. Since `Port extends Disposable`, an extension could stop a certain port forwarding by calling `Dispose`. So if a config file changes, a resolver could potentially react to it and stop forwarding the current ports and begin forwarding the new ports.

I don't agree with the argument that `PortInformation.toForward` is somehow leading to a smaller or less expanded API compared to adding `vscode.workspace.forwardPort(...): Port`. IMHO they are both API expansions, with different characteristics.

---

Finally, I think `candidates: Port[]` is typed in a funny way in this proposal. A resolver would need to return `Port` objects as candidates, but what does it mean to call `dispose()` on such candidates?

fara-nak replied:
just FYI, today we let the user to choose port number and the name they desire, and they can dispose the port anytime they want.

alexr00 replied:
@fara-nak thanks for taking a look! Choosing a name and port number will be exposed through the UI.

mjbvz commented:
@alexr00 Makes sense. I agree we can change our internal TunnelService to be more flexible when we need to support the embedder case as well

roblourens commented on src/vs/vscode.proposed.d.ts:
```diff
@@ -33,7 +33,38 @@ declare module 'vscode' {
-	export type ResolverResult = ResolvedAuthority & ResolvedOptions;
+	export interface Port extends Disposable {
+		remotePort: number;
+		localPort?: number;
```
Actually I think this should include remoteAddress (and possibly localAddress) too, not sure if that has been brought up.

alexr00 commented:
Some notes about the changes just pushed:
 - Keeping `candidatets` because we don't want to automatically forward ports just because the extension knows they're useful.
- Keeping `published` because it isn't the same as forwarded and we want to differentiate. Full discussion here: https://github.com/microsoft/vscode-remote-release/issues/1809
- Changing from `toForward` to a `workspace.forwardPort` so that extensions that have forwarded ports in settings can respond to settings changes if they want. Also just gives more flexibility.
- Separate interface for `Tunnel` and `Port` so that `candidates` and `published` make more sense.
- Added `localAddress` to `Port` and `Tunnel`.
- Removed `restorePorts` with the expectation that it becomes a setting instead.

roblourens commented on src/vs/vscode.proposed.d.ts:
```diff
@@ -33,7 +33,33 @@ declare module 'vscode' {
 
-	export type ResolverResult = ResolvedAuthority & ResolvedOptions;
+	export interface Tunnel extends TunnelDescriptor, Disposable {
+		localAddress: Uri;
```
Are these really URIs? Can they have a scheme, path, query params, etc? I think a string would do.

alexr00 replied:
A string will do, but a lot of our API takes URIs so I thought that they may often get converted into a URI anyway.

roblourens replied:
It will be like if you want to create an http url based off this info, but I don't think it makes sense for this interface to declare that the address is an http-URI, it could also be accessed in other ways.

alexr00 replied:
Changed!

alexr00 commented:
Updated again to incorporate further feedback. I think the current state will cover all known use cases, and no longer has the containers specific "published". Instead, I've changed it to detected so that if SSH, for example, wants to find list all open SSH tunnels and show them it can us the same `detected` API for it.

roblourens commented on src/vs/vscode.proposed.d.ts:
```diff
@@ -33,7 +33,31 @@ declare module 'vscode' {
+		 * The localAddress should be the complete local address for connecting to the port. Tunnels provided through
+		 * detected are read-only from the forwarded ports UI.
+		 */
+		detected?: { remotePort: number, localAddress: string }[];
```
`complete local address for connecting to the port` like a URL or just `localhost:1234`?

alexr00 replied:
just `localhost:1234`, I'll add an example to the comment.

alexr00 commented:
Thank you for all the feedback! Proposal is merged in to vscode.proposed.d.ts and will continue to evolve as it's used.
